libvips/foreign/nsgifload.c

```diff
@@ -457,30 +421,34 @@ vips_foreign_load_nsgif_generate( VipsRegion *or,
 	for( y = 0; y < r->height; y++ ) {
 		/* The page for this output line, and the line number in page.
 		 */
-		int page = (r->top + y) / gif->anim->height + gif->page;
-		int line = (r->top + y) % gif->anim->height;
+		const nsgif_info_t *info = nsgif_get_info( gif->anim );
+		int page = (r->top + y) / info->height + gif->page;
+		int line = (r->top + y) % info->height;
+		nsgif_bitmap_t *bitmap;
 
-		gif_result result;
+		nsgif_error result;
 		VipsPel *p, *q;
 
-		g_assert( line >= 0 && line < gif->anim->height );
+		g_assert( line >= 0 && line < info->height );
 		g_assert( page >= 0 && page < gif->frame_count_displayable );
 
-		if( gif->anim->decoded_frame != page ) {
-			result = gif_decode_frame( gif->anim, page ); 
-			VIPS_DEBUG_MSG( "  gif_decode_frame(%d) = %d\n", 
+//
+// I don't really understand the concept of pages
```

Comment:
GIFs are represented as very tall, thin images, with the frames laid out vertically like a strip of film.

The `_generate()` function has to fill the pixels in `or->valid`. The coordinates are in the output image space, so `or->valid.top / page_height` is the frame number, and `% page_height` is the y in this frame.

Calls to `_generate()` are guaranteed to be non-overlapping and to be in strictly sequential order (the `sequential` that `_nsgif_load()` runs upstream of generate makes this guarantee). The height of each request is unknown, so requests can potentially need several frames to fill.

The algorithm in generate is therefore:

1. For each line in the output:
2. Find the frame number and line number within that frame.
3. Is the currently loaded GIF frame the right one? If not, load the next frame.
4. Copy one line from the decoded frame into the output region.

You could do the loop in rects rather than a line at a time, but line at a time is very simple and I doubt if the speed difference would be measureable.

Looking at `nsgif_frame_decode()`, you have step 3. above in there already, so I agree, this `if()` could go, but it's handy for debugging (eg. the `print_animation()` call below) to be able to see when new frames are decoded, so I'd be inclined to keep it.

This means the `bitmap` return must be kept between `_generate()` calls, so it'll need to be a member of `VipsForeignLoadNsgif`.

Since `anim` is now opaque, `VipsForeignLoadNsgif` will need `int decoded_frame;`, and it'll need updating after the call to `nsgif_frame_decode`.

Alternatively, just call `nsgif_frame_decode()` for every scanline and get rid of `print_animation`. It's no longer very useful.

Perhaps libnsgif could have a debug method that pretty-printed the decoder state? I'd find it useful for troubleshooting tricky GIFs.